oap-server/server-configuration/configuration-etcd/src/main/java/org/apache/skywalking/oap/server/configuration/etcd/EtcdConfigWatcherRegister.java

```diff
@@ -0,0 +1,148 @@
+        responsePromiseByKey = new ConcurrentHashMap<>();
+    }
+
+    @Override public ConfigTable readConfig(Set<String> keys) {
+        removeUninterestedKeys(keys);
+        registerKeyListeners(keys);
+        final ConfigTable table = new ConfigTable();
+
+        for (Map.Entry<String, Optional<String>> entry : configItemKeyedByName.entrySet()) {
+            final String key = entry.getKey();
+            final Optional<String> value = entry.getValue();
+
+            if (value.isPresent()) {
+                table.add(new ConfigTable.ConfigItem(key, value.get()));
+            } else {
+                table.add(new ConfigTable.ConfigItem(key, null));
+            }
+        }
+
+        return table;
+    }
+
+    private void registerKeyListeners(final Set<String> keys) {
+        for (final String key : keys) {
+            String dataId = "/" + settings.getGroup() + "/" + key;
+            if (listenersByKey.containsKey(dataId)) {
+                continue;
+            }
+
+            listenersByKey.putIfAbsent(dataId, p -> {
+                onDataValueChanged(p);
+            });
+
+            try {
+                EtcdResponsePromise<EtcdKeysResponse> responsePromise = client.get(dataId).waitForChange().send();
+                responsePromise.addListener(listenersByKey.get(dataId));
+                responsePromiseByKey.putIfAbsent(dataId, responsePromise);
+
+                // the key is newly added, read the config for the first time
+                EtcdResponsePromise<EtcdKeysResponse> promise = client.get(dataId).send();
+                onDataValueChanged(promise);
+            } catch (Exception e) {
+                throw new EtcdConfigException("wait for etcd value change fail", e);
+            }
+        }
+    }
+
+    private void removeUninterestedKeys(final Set<String> interestedKeys) {
+        final Set<String> uninterestedKeys = new HashSet<>(listenersByKey.keySet());
+        uninterestedKeys.removeAll(interestedKeys);
+
+        uninterestedKeys.forEach(k -> {
+            final ResponsePromise.IsSimplePromiseResponseHandler listener = listenersByKey.remove(k);
+            if (listener != null) {
+                responsePromiseByKey.remove(k).removeListener(listener);
+            }
+        });
+    }
+
+    private void onDataValueChanged(ResponsePromise<EtcdKeysResponse> promise) {
+        try {
+            EtcdKeysResponse.EtcdNode node = promise.get().getNode();
+            String dataId = getRealKey(node.getKey(), settings.getGroup());
+            String value = node.getValue();
+            if (logger.isInfoEnabled()) {
+                logger.info("Nacos config changed: {}: {}", dataId, node.getValue());
+            }
+
+            configItemKeyedByName.put(dataId, Optional.ofNullable(value));
+        } catch (Exception e) {
+            if (e instanceof EtcdException) {
+                if (EtcdErrorCode.KeyNotFound == ((EtcdException)e).errorCode) {
+                    return;
```

Comment:
simply ignore `EtcdErrorCode.KeyNotFound`? do we need something like `configItemKeyedByName.put(dataId, Optional.empty());`

Reply:
As exception is so much that it doesn't say that we need put null to the table. we do this only we found that no value found. This is my idea. what do you think ? @kezhenxu94

Reply:
> As exception is so much that it doesn't say that we need put null to the table. we do this only we found that no value found. This is my idea. what do you think ? @kezhenxu94

I'm not sure, I was thinking that etcd would throw such exception (`KeyNotFound`) when we are getting a deleted key, the exception name `KeyNotFound` made me think that, it's just my personal thought and may be wrong, please ignore if you had confirmed, thanks :)

Reply:
@wayilau Ping me if you are sure and I'll mark this as resolved :)

Reply:
@kezhenxu94 You are right, we need set empty for this when key not found. It is say that key is not exists for the first time or be deleted in the later. For other exception, we just throw, do not change the table.

Reply:
@kezhenxu94 You are right, we shoud add empty when key was first time created or when it is deleted in later time. I have changed this.